python/pywatchman/pybser.py

```diff
@@ -213,45 +213,106 @@ def _bunser_string(buf, pos):
     return (str_val, pos + str_len)
 
 
-def _bunser_array(buf, pos):
+def _bunser_array(buf, pos, mutable=True):
     arr_len, pos = _bunser_int(buf, pos + 1)
     arr = []
     for i in range(arr_len):
-        arr_item, pos = _bser_loads_recursive(buf, pos)
+        arr_item, pos = _bser_loads_recursive(buf, pos, mutable)
         arr.append(arr_item)
+
+    if not mutable:
+      arr = tuple(arr)
+
     return arr, pos
 
 
-def _bunser_object(buf, pos):
+# This is a quack-alike with the bserObjectType in bser.c
+# It provides by getattr accessors and getitem for both index
+# and name.
+class _BunserDict(object):
```

Comment:
My first pass at this just used dict and added the getattr out of laziness, but then I realized that I should make it more like the thing we're exporting from the native C extension to minimize the chance they pybser behaves differently from the native extension.  I'm still pretty new to the python internals so I don't know exactly how close or how far from that goal we are here :-)